pkg/development/devcontext.go

```diff
@@ -0,0 +1,303 @@
+				Context: tuple.String(tpl),
+			})
+			continue
+		}
+
+		err := validateTupleWrite(ctx, tpl, nsm, revision)
+		if err != nil {
+			devErr, wireErr := DistinguishGraphError(ctx, err, v0.DeveloperError_RELATIONSHIP, 0, 0, tuple.String(tpl))
+			if devErr != nil {
+				devErrors = append(devErrors, devErr)
+				continue
+			}
+
+			return decimal.NewFromInt(0), devErrors, wireErr
+		}
+
+		updates = append(updates, &v1.RelationshipUpdate{
+			Operation:    v1.RelationshipUpdate_OPERATION_TOUCH,
+			Relationship: tuple.MustToRelationship(tpl),
+		})
+	}
+
+	revision, err := ds.WriteTuples(ctx, nil, updates)
+	return revision, devErrors, err
+}
+
+func loadNamespaces(
+	ctx context.Context,
+	namespaces []*v0.NamespaceDefinition,
+	nsm namespace.Manager,
+	ds datastore.Datastore,
+) ([]*v0.DeveloperError, decimal.Decimal, error) {
+	errors := make([]*v0.DeveloperError, 0, len(namespaces))
+	var lastRevision decimal.Decimal
+	for _, nsDef := range namespaces {
+		ts, terr := namespace.BuildNamespaceTypeSystemForDefs(nsDef, namespaces)
+		if terr != nil {
+			return errors, lastRevision, terr
+		}
+
+		tverr := ts.Validate(ctx)
+		if tverr == nil {
+			var err error
+			lastRevision, err = ds.WriteNamespace(ctx, nsDef)
+			if err != nil {
+				return errors, lastRevision, err
+			}
+			continue
+		}
+
+		errors = append(errors, &v0.DeveloperError{
+			Message: tverr.Error(),
+			Kind:    v0.DeveloperError_SCHEMA_ISSUE,
+			Source:  v0.DeveloperError_SCHEMA,
+			Context: nsDef.Name,
+		})
+	}
+
+	return errors, lastRevision, nil
+}
+
+// DistinguishGraphError turns an error from a dispatch call into either a user-facing
+// DeveloperError or an internal error, based on the error raised by the dispatcher.
+func DistinguishGraphError(ctx context.Context, dispatchError error, source v0.DeveloperError_Source, line uint32, column uint32, context string) (*v0.DeveloperError, error) {
+	var nsNotFoundError sharederrors.UnknownNamespaceError
+	var relNotFoundError sharederrors.UnknownRelationError
+
+	if errors.As(dispatchError, &nsNotFoundError) {
+		return &v0.DeveloperError{
+			Message: dispatchError.Error(),
+			Source:  source,
+			Kind:    v0.DeveloperError_UNKNOWN_OBJECT_TYPE,
+			Line:    line,
+			Column:  column,
+			Context: context,
+		}, nil
+	}
+
+	if errors.As(dispatchError, &relNotFoundError) {
+		return &v0.DeveloperError{
+			Message: dispatchError.Error(),
+			Source:  source,
+			Kind:    v0.DeveloperError_UNKNOWN_RELATION,
+			Line:    line,
+			Column:  column,
+			Context: context,
+		}, nil
+	}
+
+	var ire invalidRelationError
+	if errors.As(dispatchError, &ire) {
+		return &v0.DeveloperError{
+			Message: dispatchError.Error(),
+			Source:  source,
+			Kind:    v0.DeveloperError_UNKNOWN_RELATION,
+			Line:    line,
+			Column:  column,
+			Context: context,
+		}, nil
+	}
+
+	return nil, rewriteACLError(ctx, dispatchError)
+}
+
+func rewriteACLError(ctx context.Context, err error) error {
```

Comment:
this is copied from `internal/services/v0`, right?

should this just be public and shared?

Reply:
Yes, and I don't think so. Once v0 is gone, it'll only be needed in here, so I decided to just copy it